debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModel_V2.java

```diff
@@ -67,4 +69,40 @@ public UDATA getTotalFootprintInBytesWithHeader(J9IndexableObjectPointer arrayPt
 		UDATA totalFootprint = spineSize.add(externalArrayletSize);
 		return totalFootprint;
 	}
+
+	/**
+	 * Determine the validity of the data address belonging to arrayPtr.
+	 *
+	 * @param arrayPtr array object who's data address validity we are checking
+	 * @throws CorruptDataException if there's a problem accessing the indexable object dataAddr field
+	 * @return true if the data address of arrayPtr is valid, false otherwise
+	 */
+	@Override
+	public boolean hasCorrectDataAddrPointer(J9IndexableObjectPointer arrayPtr) throws CorruptDataException
+	{
+		boolean hasCorrectDataAddrPointer;
+		try {
+			UDATA dataSizeInBytes = getDataSizeInBytes(arrayPtr);
+			VoidPointer dataAddr = J9IndexableObjectHelper.getDataAddrForIndexable(arrayPtr);
+			boolean isValidDataAddrForDoubleMappedObject = isIndexableObjectDoubleMapped(dataAddr, dataSizeInBytes);
+
+			if (dataSizeInBytes.isZero()) {
+				VoidPointer discontiguousDataAddr = VoidPointer.cast(arrayPtr.addOffset(J9IndexableObjectHelper.discontiguousHeaderSize()));
+				hasCorrectDataAddrPointer = (dataAddr.isNull() || dataAddr.equals(discontiguousDataAddr));
+			} else if (dataSizeInBytes.lt(arrayletLeafSize)) {
+				VoidPointer contiguousDataAddr = VoidPointer.cast(arrayPtr.addOffset(J9IndexableObjectHelper.contiguousHeaderSize()));
+				hasCorrectDataAddrPointer = dataAddr.equals(contiguousDataAddr);
+			} else {
+				if (enableVirtualLargeObjectHeap || enableDoubleMapping) {
+					hasCorrectDataAddrPointer = isValidDataAddrForDoubleMappedObject;
```

Comment:
probably it's cleaner if we just call  !isAddressWithinHeap (since isIndexableObjectDoubleMapped also checks that size > leaf size, which we already know in thise else branch)

in future, we could also be more specific and check that addr is within off-heap range (if it's easy to fetch)